plugins/user_mapping/src/local_file.cpp

```diff
@@ -0,0 +1,173 @@
+#include "interface.h"
+
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <filesystem>
+#include <fstream>
+#include <mutex>
+#include <optional>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+#include <spdlog/spdlog.h>
+
+namespace
+{
+	struct user_profile
+	{
+		std::string irods_user_name;
+		nlohmann::json attributes;
+	};
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+	std::string file_path;
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+	std::vector<user_profile> profile_list;
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables, cert-err58-cpp)
+	std::shared_mutex list_mutex;
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+	std::mutex update_mutex;
+
+	auto init(const nlohmann::json& _config) -> void
+	{
+		const auto path{_config.find("file_path")};
+		if (path == std::end(_config)) {
+			// handle bad case
+			spdlog::error("{}: Unable to find [file_path] in provided config.", __func__);
+		}
+
+		// Save file string
+		file_path = path->get<std::string>();
+	} // init
+
+	auto update() -> void
+	{
+		static std::filesystem::file_time_type last_file_path_write;
+
+		// If there have been no changes to file_path, there is no work to do
+		if (std::filesystem::last_write_time(file_path) == last_file_path_write) {
+			return;
+		}
+
+		std::ifstream file{file_path};
+		spdlog::debug("{}: Opened file [{}]", __func__, file_path);
+
+		auto json_data{nlohmann::json::parse(file)};
+
+		// Create temp list to be swapped in later
+		std::vector<user_profile> temp_list;
+		temp_list.reserve(json_data.size());
+
+		// Process into list:
+		auto base_iter{json_data.items()};
+		std::transform(
+			std::begin(base_iter),
+			std::end(base_iter),
+			std::back_inserter(temp_list),
+			[](const auto& _iter) -> user_profile {
+				return {.irods_user_name = _iter.key(), .attributes = _iter.value()};
+			});
+
+		// Move new item list to old one
+		if (std::unique_lock update_profile_list_lock{list_mutex, std::try_to_lock}; update_profile_list_lock) {
```

Comment:
I believe you can just block here since the update mutex only allows one thread to reach this logic.

Will need to think about this a little more though.

Reply:
Here's what I could think of for both:

`std::try_to_lock`:
- Pros
  - Lets the current thread's work progress quickly
- Cons
  - `update()`'s work could be wasted if locking fails
  - Single write, multi read could mean `update()`'s write gets delayed until plugin traffic gets less busy.
    - This applies to both methods of lock, the effects are different, though. In this case, work may be thrown away multiple times.

`std::lock`:
- Pros
  - Ensures that `update()`'s work is ensured to not be wasted
- Cons
  - Could be blocked for a long time if `std::shared_mutex` continues to lock.
    - Think mapping with a large set (longer thread execution), and multiple auths at once.

---

Interesting detail of `std::try_to_lock` I just learned:

> An implementation may fail to obtain the lock even if it is not held by any other thread.

Seen in C++ Working Draft. 33.6.4.2.1 [thread.mutex.requirements.mutex.general]

https://github.com/irods/irods_client_http_api/blob/e977b247714d7b244c7dbe6a98dab1c2dd1585bd/plugins/user_mapping/src/local_file.cpp#L91-L94

This should not be an issue for the lock around the call to `update()`, as there are multiple chances to try again to call update.

Reply:
After discussion, we decided to use the block lock here. Since there is already a lock on the `update()` call, a blocking lock within the call should be fine.

Reply:
Done.